[rom_ctrl] Make it possible to configure the rom to take 2 cycles

rswarbrick opened the pull request:
The first commit in this PR adds a configurable extra flop to the ROM output, which means that anyone using `prim_rom` will see it take 2 cycles to respond. The rest of the commits (surprisingly fiddly!) are teaching `rom_ctrl` to allow this configuration. 

Amusingly, the 4th commit actually makes things a bit simpler: I suspect this shows that I wasn't quite as competent a few years ago!

rswarbrick committed:
[prim,rtl] Add a Latency parameter to prim_rom

Also document the module a bit more fully and change the (clearly
non-negative) parameters to be "int unsigned", which should make the
intent a bit more obvious.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[prim,rtl] Add a Latency parameter to prim_rom_adv

This is analogous to the calculation in prim_rom and controls the
latency for the rvalid_o port.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[rom_ctrl,rtl] Allow rom_ctrl_scrambled_rom with Latency 2

The only work needed is that the calculated keystream needs delaying
by a cycle to match the rdata_scr signal.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[rom_ctrl,rtl] Move a localparam in rom_ctrl_fsm

No functional change, but it allows a simpler description of a couple
of port widths.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[rom_ctrl,rtl] Add simple consistency check for kmac_data_o.last

This wasn't actually true for one of my attempteded versions of the
design, and is very easy to check.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[rom_ctrl,rtl] Pass a TwoCycleRom parameter from top of rom_ctrl

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[rom_ctrl,rtl] Teach rom_ctrl_mux about 2-cycle version of ROM

The mux passes rom_rvalid_i back to the bus (in bus_rvalid_o), so
needs to know the ROM latency because it needs to know exactly when
the selection switches between the FSM and the bus.

Conveniently, we have both registered versions of sel_bus_i and have
checked whether each is Mubi4True anyway.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

andreaskurth commented:
Force-pushed to implement @davidschrammel's suggestion -- thanks!

Razer6 commented:
> edit: In today's meeting it was suggested that we change this entirely and make (and use) top-specific sim configs. This would also make earlgrey tests run faster since that top only requires one of them instead of the 32(?)kB cfg.

My proposal is to make those changes in a follow-up PR, I can take care of that to get that merged in first. @vogelpi since you have maintainer rights, can you rebase that PR to resolve the conflicts?

Razer6 reviewed (changes requested):
I have a few requests. In general what I think is we should not tie that to a 2-cycle ROM but instead make the ROM latency configurable. The implementation for that is already 90% I think, let's fix the remaining to have a generic and scalable implementation - what if the next comes and says I have a 3 cycle ROM.

vogelpi committed:
[rom_ctrl,rtl] Drop address of returned data from rom_ctrl_counter

The reason to look at this is that I want to parameterise the response
latency of the ROM itself. Since the counter doesn't contain the ROM,
it feels wrong for it to know or care about the ROM's responses.

Working through this carefully has made me realise that I didn't
design it very cleanly in the first place! The slightly rejigged
version here should also be a bit simpler to reason about.

Co-authored-by: Rupert Swarbrick <rswarbrick@lowrisc.org>
Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[rom_ctrl,data] Expose the new TwoCycleRom parameter in hjson

Co-authored-by: Rupert Swarbrick <rswarbrick@lowrisc.org>
Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi committed:
[rom_ctrl,rtl] Add configurable ROM latency to rom_ctrl_fsm

This makes the module marginally more complicated, mainly because we
have to be a little more careful about storing ROM responses if KMAC
changes its mind about being ready after we have sent the requests.

Co-authored-by: Rupert Swarbrick <rswarbrick@lowrisc.org>
Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

vogelpi commented on hw/top_darjeeling/data/top_darjeeling.hjson:
```diff
@@ -887,7 +887,8 @@
       param_decl: {
-        SecDisableScrambling: "1'b0"
+        SecDisableScrambling: "1'b0",
+        TwoCycleRom: "1'b1"
```
I've now disabled this again for Darjeeling (dropped the last commit from the PR).

vogelpi committed:
[rom_ctrl/dv] Add build options for recently added RTL parameters

Signed-off-by: Pirmin Vogel <vogelpi@lowrisc.org>

davidschrammel commented:
I think it is worrying that there's DV problems that we're kind of hiding by testing a different configuration that the one that is defined in the top hjson files.

Can we capture the DV issues in a github issue and a TODO comment with the issue number?
Maybe also another issue that describes that the block level sim cfg can drift apart from the toplevel configurations?
